CMake + Windows support + CI/CD

hugsy opened the pull request:
Hi,

Like I mentioned in #14 here is my current PR for all the changes I brought to your great project:
 1. `cmake` is now the build system generator to replace `autoconf`, simply because `cmake` works universally. Also started adding (some of the) test rules, because to port them all requires a lot of changes and I wasn't sure you'd be comfortable to adding that many changes all at once.
 1. made the library fully Windows compatible (which is the main reason for switching to `cmake`)
 1. integration to GH Actions for CI + CD purposes: it will automatically test the bins generated (only `asmline` and `run` so far) and if successful will generate artifacts for both windows (header, lib, dll and pdb) and linux (headers and .so) : see https://github.com/hugsy/AssemblyLine/actions/runs/1894958934 for an example . Pretty sure `macos-latest` can be added to generate artifacts for macos too without trouble.
 1. fixed a logic bug in `tokenizer.c` (with the use of `strtok`) found by sanitizer

Great project, super useful! The only useful alternative was keystone-engine, but this is *way* lighter and easier to embed.

Cheers,

Commit:
moved to cmake + windows support

Commit:
added `asmline`

Commit:
get rid of the `autoconf` stuff

Commit:
added actions

Commit:
[ci] fixed path

hugsy committed:
Update build.yml

Commit:
added tools/cmakelist

Commit:
added test/cmakelist

Commit:
asd

Commit:
ported changes up to 0xADE1A1DE/AssemblyLine@be2fab3

Commit:
fixed bug in tokenizer.c (bad use of `strtok_s`)

Commit:
test stuff

Commit:
ci fix

Commit:
ci fix

Commit:
[ci win] switch msbuild actions

Commit:
last touches

Commit:
last touches

Commit:
added config.h.in

hugsy committed:
Update README.md

dderjoel commented:
Hi, 
that looks wonderful! Thank you very much!
We are happy to move to `cmake` if we can get included:
- `man` pages (does windows have an equivalent?)
- `make install` (or at least some documentation on how to install, and what the dependencies are, we are rather un-familiar with `cmake`, (`make install` does not work, and `cmake --install . --prefix /tmp` )
- `make check` to go through the test directory and check if `asmline` emits the same binary code as `nasm` would.
  - nice2have: the `tap` files, which are typically used during development to check where / which instruction is different to `nasm`'s
  - nice2have: the `eaf` files, which are test-cases, where we 'expect assmeblyline to fail' -- on invalid input (rather than silently ignoring/assembling something)
- `pc` installation file, for dependent projects to be able to specify required versions

hugsy commented:
👋

AFAIK the closest that Windows has to a `man` page is `Get-Help` command for Powershell so we can add a PS1 script that would be the equivalent of `asmline` and hit the dll directly.
Pretty much everything (like the `make check` and `make install`) that is/was on autoconf can be ported to cmake, it is made for that. I didn't do an exhaustive job in case you weren't interested in merging it :D 

I'm not sure what the `pc` is for though.

Commit:
added `install` directives to `cmake` files

Commit:
Merge branch 'main' of https://github.com/hugsy/AssemblyLine

Commit:
- Added cmake `test` rules
- By default, don't build tests

Commit:
updated version string from cmake

hugsy commented:
- Install added by 33f6f2d
  -  `cmake --install . --prefix /whereever/assemblyline`  will work
- Test added by 070cd98
  - Simply do `cmake .. && make && cd test && ctest`
- Man install added by d46c229

Commit:
added `man` installation rule

Commit:
added depencies

Commit:
CI fixes

Commit:
added windows tests for GH Actions

dderjoel commented:
> I'm not sure what the `pc` is for though.

The high level idea is that you can require versions of dependent libraries in your third party library/tool. 
So I have another tool `foo` and it needs `libassemblyline` but requires it to be version `1.2.0` or later, because it uses its newest features.
Then, I can specify in my project that as a dependency at compile-time. Using `autotools` that would be something like
```
PKG_CHECK_MODULES([LIBASSEMBLYLINE], [assemblyline >= 1.2.0])
```
more info [here](https://people.freedesktop.org/~dbn/pkg-config-guide.html)

Do you know if there a `cmake` equivalent / alternative (preferably, I could still use it from `foo`, based on `autotools`, but that is probably rather tricky...)?

> AFAIK the closest that Windows has to a `man` page is `Get-Help` command for Powershell so we can add a PS1 script that would be the equivalent of `asmline` and hit the dll directly.

So, you mean an alias from `Get-Help asmline` to `asmline --help` ? But that would not necessarily get the information on how to use `libassemblyline.dll`, right?

re install:
- can we specify, that only the api-header is copied `assemblyline.h` and not all the internal ones, as they should not be considered stable?
- the installation copies to `lib/libAssemblyLine.so`. As unix is case-sensitive, we need to change that all lower-case. Secondly, it does not create the confusing, but necessary links.
    ```
    $ ls -l /usr/lib/libassem*
    lrwxrwxrwx 1 root root    24 Feb 28 12:50 /usr/lib/libassemblyline.so -> libassemblyline.so.1.2.1
    lrwxrwxrwx 1 root root    24 Feb 28 12:50 /usr/lib/libassemblyline.so.1 -> libassemblyline.so.1.2.1
    -rwxr-xr-x 1 root root 67416 Feb 28 12:50 /usr/lib/libassemblyline.so.1.2.1
    ```
  which are needed for the dynamic linker to link against the correct `current` version. This is described in the `autotools` [here](https://autotools.info/libtool/version.html). As far as I could see, that is now in `project(AssemblyLine LANGUAGES C VERSION 1.2.0)` (d46c2298fe3fd1082bb40753a32a60f2cec80731) -- Does that refer to the `Major.Minor.Patch` (semver) or to the `Current.Age.Revision`? And does cmake generate the 'correct other' ?

dderjoel commented:
> Test added by [070cd98](https://github.com/0xADE1A1DE/AssemblyLine/commit/070cd984413dcaffe40b7d25371d66c0b427377c)
> 
>     * Simply do `cmake .. && make && cd test && ctest`

well, as also your commit message states, tests are not built by default. Looking at the CI yaml, I tried `CMAKE_FLAGS='-DBUILD_TEST=ON' cmake .. && make` but that also does not create the test-dir. :/

hugsy commented:
Hey!

> The high level idea is that you can require versions of dependent libraries in your third party library/tool.

Ah yes no problem, I forgot about the `pkg-config` file (my focus was initially for Windows :smile:) . But here again, cmake can generate one very easily. I will add it (update: added by f2f7301)

> Do you know if there a cmake equivalent / alternative (preferably, I could still use it from foo, based on autotools, but that is probably rather tricky...)?

cmake is really flexible. What you seem to describe is called a custom target in the cmake lingo and you can find tons of resources on how to create one (for instance [this one](https://dev.to/iblancasa/learning-cmake-3-understanding-addcustomcommand-and-addcustomtarget-43gp) doesn't look too bad for intro).

> can we specify, that only the api-header is copied assemblyline.h and not all the internal ones, as they should not be considered stable?

Totally, cmake uses a system a private/public interface allowing to define what's meant to be public or not. But if we wanna go simple for this project, we can do that by simply updating the line to mention to install only `assemblyline.h`

https://github.com/hugsy/AssemblyLine/blob/a15ca3f4a0dd0d82bfad9e65c972f80d24d14717/src/CMakeLists.txt#L88

> the installation copies to lib/libAssemblyLine.so. As unix is case-sensitive, we need to change that all lower-case. Secondly, it does not create the confusing, but necessary links.

It's all feasible easily, by creating a [POST_BUILD custom command](https://cmake.org/cmake/help/latest/command/add_custom_command.html#build-events), we can create as many symlinks we want.

> Does that refer to the Major.Minor.Patch (semver) 

Yes, those will be auto generated by cmake and exposed to the build system (which can also be exposed to the source being compiled): see https://cmake.org/cmake/help/latest/command/project.html#options

> The high level idea is that you can require versions of dependent libraries in your third party library/tool.

Yes it is created, you can verify on Github Actions: see https://github.com/hugsy/AssemblyLine/runs/5368746701?check_suite_focus=true#step:11:100249

Hope this helps.

Cheers

Commit:
let cmake generate the pkgconfig file

hugsy committed:
[ci] added pkgconfig file to linux artifact generation

dderjoel commented:
Hi,

- generating `pkg-config`: nice! We need this file to be copied to `INSTALLDIR/lib/pkgconfig/assemblyline.pc`(now my dependent-autotools-project can do what it used to do - wonderful)
- custom targets. I'm not sure, if that is what I mean. looking a bit more into `cmake` documentation, I think I want something like `target_link_libraries(MyLib -lassemblyline)` if I have another project, which needs to be linked to `assemblyline`. However, I want to specify that I need to link to it only if it has a specific (minimal) version. Something like `target_link_libraries(MyLib -lassemblyline>1.2.0)` which should error out, if only version 1.1.0 of assemblyline is installed. (Autotools use that `pkg-config` file for that).
- 

> But if we wanna go simple for this project, we can do that by simply updating the line to mention to install only `assemblyline.h`

yes, lets do that and only install `assemblyline.h`

> It's all feasible easily, by creating a [POST_BUILD custom command](https://cmake.org/cmake/help/latest/command/add_custom_command.html#build-events), we can create as many symlinks we want.

Lets not create them ourselves.
This seems to be what [this thread is about](https://stackoverflow.com/questions/42183903/cmake-wont-create-so-version-and-soversion-symlinks). I see that in the `src/CMakeLists.txt` the library is already `SHARED`, but the `set_target_properties` is not used. When I try to just copy-n-paste (replacing `${PLUGIN_NAME}` with `AssemblyLine` that in there, the installation creates the following links:
```
# joel in /tmp/hugsy/build on git:main x [10:46:49] 
$ ll ../install/lib 
drwxr-xr-x 2 joel joel   100 Mar  2 10:46 ./
drwxr-xr-x 6 joel joel   120 Mar  2 10:46 ../
lrwxrwxrwx 1 joel joel    25 Mar  2 10:46 libAssemblyLine.so -> libAssemblyLine.so.PREFIX*
-rwxr-xr-x 1 joel joel 82240 Mar  2 10:46 libAssemblyLine.so...*
lrwxrwxrwx 1 joel joel    21 Mar  2 10:46 libAssemblyLine.so.PREFIX -> libAssemblyLine.so...*
```
So I'd assume those version Variables are not populated correctly. Can you advice how to set them from the `PROJECT_VERSION` defined in the root-`CMakeLists.txt`:6 ? Important here, is that if calculates them correctly (-> to current/age/revision as mentioned earlier (I remember finding some formulas somewhere, with disgusting m4-replacemet macros etc)).

And lastly, for the tests (which env-vars do I need to let `cmake` generate the `test`-dir?), I can't see where `asmline -p` is called (that's what it looks like on the CI-run). Also, is this output compared for correctness against `nasm`?

hugsy commented:
> Something like target_link_libraries(MyLib -lassemblyline>1.2.0) which should error out, if only version 1.1.0 of assemblyline is installed. (Autotools use that pkg-config file for that).

You don't really have to worry about that with `cmake`, as long as yours are correctly created: if your project is used as an internal submodule (like I do now), all the caller has to do is check `WhateverLib_VERSION` constant.
If called externally, [`find_package`](https://cmake.org/cmake/help/v3.23/command/find_package.html) does exactly that (including from `pkgconfig`).

> So I'd assume those version Variables are not populated correctly. Can you advice how to set them from the PROJECT_VERSION defined in the root-CMakeLists.txt:6 ?

You're correct, using `set_target_properties( ... SOVERSION ...)` is the way to go. Here something like htis is enough:
```cmake
set_target_properties(AssemblyLine PROPERTIES 
                                 VERSION ${PROJECT_VERSION}
                                 SOVERSION ${PROJECT_VERSION_MAJOR})
```

Because we set the `VERSION` of `project()`  correctly.

```bash
$ cmake ~/code/AssemblyLine -DBUILD_TEST=ON && make -j4 && cmake --install . --prefix /tmp/assemblyline
-- The C compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
[...]
$ ls -l /tmp/assemblyline/lib
lrwxrwxrwx  20 hugsy  1 Mar 19:20 libAssemblyLine.so -> libAssemblyLine.so.1
lrwxrwxrwx  24 hugsy  1 Mar 19:20 libAssemblyLine.so.1 -> libAssemblyLine.so.1.2.0
.rw-r--r-- 83k hugsy  1 Mar 19:20 libAssemblyLine.so.1.2.0
```

Is that what you want?

> And lastly, for the tests (which env-vars do I need to let cmake generate the test-dir?), I can't see where asmline -p is called (that's what it looks like on the CI-run). 

That's really up to you (and that's kindda why it's hard for me to do stuff, because you may not like them the way I do, so I go step-by-step). But with the implementation I made, yes `cmake` created the stub to make your build system create the test directory so you have nothing to do.

> Also, is this output compared for correctness against nasm?

Currently no, how do you want this to be done exactly? You can use `ctest` and match the output using `cmake`, I like this approach but at the end it's up to you. [The cmake tutorial](https://cmake.org/cmake/help/v3.23/guide/tutorial/Installing%20and%20Testing.html#testing-support) details well how you can create an intermediate function to match the output from a regexp.

If you confirm the target property like I mention above, let me know, and I'll send the modification (along with the rule for install of `assemblyline.h` ).

Cheers

dderjoel commented:
> You don't really have to worry about that with `cmake`, as long as yours are correctly created:

Ah yes, you're right. Not for now. If I would be converting my dependent project to `cmake`, then I'd need to worry. 

> Is that what you want?

Yes, That is Perfect!

> Currently no, how do you want this to be done exactly?

Hmmm. So what we *need* is assemblyline to create _correct_ assembly. For that, we usually check against `nasm` on the same input. But, we have some constructs, which are incompatible (i.e. assemblyline emits semantically equivalent code, which differs from `nasm`) For the general case, (equivalent to `nasm`) we just had the test-files, assemble them with both programs, convert them to the same hex-format and run `diff`. This is the essence of the scripts [here](https://github.com/0xADE1A1DE/AssemblyLine/blob/main/test/al_nasm_compare.sh) and [here](https://github.com/0xADE1A1DE/AssemblyLine/blob/main/test/tap/compiler.sh). The former compiles the entire file and checks for equivalence, the latter uses the [TAP](https://testanything.org/)-Protocol, and checks line by line; that is slower, but easier to see, which single instruction may still be problematic.

From your link, however, I understand that one would need to write the expected output into the `CMakeLists.txt` file. The benefit of that would be, that we can check also the nasm-incompatible code. And we should be able to create such a file (to get all the existing tests sorted) with a couple scripts, I guess.

I'm also happy to introduce some frameworks to have the code tested, or to write this in `c`-files, too and / or run the code to see if the expected happens (I'm pretty sure, that would be a bit overkill for this PR).

What I don't want, is that we merge all this (your) great work at the cost of abandoning the (correctness-)testing.
@daviduwu9 , what do you think / prefer - so we (you, @daviduwu9) want to experiment with different approaches and then come back for a decision? Or does @hugsy have a solid idea to get this solved, now knowing the requirements? I'm pretty happy with whatever, at this point.


Does this 


> If you confirm the target property like I mention above, let me know, and I'll send the modification (along with the rule for install of `assemblyline.h` ).

refer to this

> Totally, cmake uses a system a private/public interface allowing to define what's meant to be public or not. But if we wanna go simple for this project, we can do that by simply updating the line to mention to install only `assemblyline.h`
> 
> https://github.com/hugsy/AssemblyLine/blob/a15ca3f4a0dd0d82bfad9e65c972f80d24d14717/src/CMakeLists.txt#L88

?

Then, I'm not sure what you want me to confirm.

dderjoel commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,93 @@
+# Install directives
+#
+install(TARGETS AssemblyLine DESTINATION lib)
+install(FILES ${HEADER_FILES} DESTINATION include)
```
```suggestion
install(FILES assemblyline.h DESTINATION include)
```

dderjoel replied:
Is that what you'd like me to confirm ? sure. Or would we want to change the variable `HEADER_FILES` instead (I don't know if/where it is used elsewhere)

hugsy commented:
Ok I get it now, changes coming soon.

Yeah we can definitely generate tests to be executed against nasm, my point was more that there's plenty of way to do so. I will propose an approach and let you guys judge.
And FTR I'm 100% in line with you, code should be tested exhaustively.

Cheers

Commit:
PR feedback
- lowercase lib name
- generate lib symlinks
- only install `assemblyline.h`

Commit:
installing `pc` file at the right spot

Commit:
Merge branch 'main' of https://github.com/hugsy/AssemblyLine

Commit:
tentative approach to replace the use of nasm
 - we let gcc generate the byte code from the assembly code to test, then at runtime, use assemblyline and compare both

hugsy commented:
I had a little bit of time so I went ahead and try a new approach for testing, let me know what you think: basically this removes the need of external tool, we let the compiler generate the bytecode, that we can compare with that of asmline lib at runtime of test. This also allows to be fully integrated with any build system (although msvc doesn't support x64 inline `__asm__`). The new file is pure C, small and easy to read (IMHO) for being modified in the future.

The source is [here](https://github.com/0xADE1A1DE/AssemblyLine/blob/76b8aca1fffc4b27291a365da0c4301393edff15/test/compile_asm.c) and it works great on my host:
<img width="277" alt="image" src="https://user-images.githubusercontent.com/590234/156315596-58fe5daf-9984-4cad-b822-46c904521d40.png">
`[...]`
<img width="203" alt="image" src="https://user-images.githubusercontent.com/590234/156315662-a812afb5-1ee9-4601-a7e2-3e3f5ea450ae.png">

Commit:
[ci] prevent test/compile_asm to be built on win x64

Commit:
flag adjustment

dderjoel commented on test/compile_asm.c:
```diff
@@ -0,0 +1,92 @@
+    size_t test_sz = __stop_mytest - __start_mytest;
+    printf("size = %lu\n", test_sz);
+
+    // prepare asmline
+    assemblyline_t al = asm_create_instance(NULL, 0);
+    if ( asm_assemble_str(al, CODE) == EXIT_FAILURE )
+        return EXIT_FAILURE;
+
+    void (*fun)(uint64_t * out, uint64_t * in0, ...) = asm_get_code(al);
+    unsigned char* c1                                = (unsigned char*)fun;
+
+    // adjust (because gcc always adds some junk instructions)
+    unsigned char* c2 = (unsigned char*)test;
+    c2 += 4;
+    test_sz -= 6;
+
+    // compare
+    size_t nb_errors = 0;
+    for ( size_t offset = 0; offset < test_sz; offset++ )
+    {
+        char valid = (c1[offset] == c2[offset]);
+        printf("[%ld] %#.02x == %#.02x ? %s\n", offset, c1[offset], c2[offset], valid ? "TRUE" : "FALSE");
+        if ( !valid )
+            nb_errors++;
+    }
+
+    printf("%lu error(s) found\n", nb_errors);
+    return nb_errors == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
```
```suggestion
  // adjust (because gcc always adds some junk instructions)
    unsigned char* c2 = (unsigned char*)test;
    c2 += 4;
    test_sz -= 6;


    // prepare asmline
    assemblyline_t al = asm_create_instance(NULL, 0);

    // put 'code' to a variable on the stack
    char c1[] = CODE;
    char *s   = c1, *savetok;
    int pos, len, result;


    char* token = strtok_r(c1, "\n", &savetok);
    do
    {
        pos         = asm_get_offset(al);
        char* start = asm_get_code(al);

        if ( asm_assemble_str(al, token) != EXIT_SUCCESS )
        {
            goto fail;
        }
        char* end = asm_get_code(al);
        while ( start < end )
        {
            if ( *start++ != *c2++ )
                goto fail;
        }
    } while ( (token = strtok_r(NULL, "\n", &savetok)) );

    result = EXIT_SUCCESS;
    goto end;

fail:
    fprintf(stderr, "failed at instruction >>%s<<\n", token);
    result = EXIT_FAILURE;

end:
    asm_destroy_instance(al);
    return result;

```

dderjoel commented:
I do like the fact to just rely on `cc`, and then compare that to our lib. I do not like, that we ~compare byte-by-byte~ compile everything *and then* compare , because we don't know which instruction is problematic.
I am thinking of using `asm_get_code` and `asm_assemble_str` in a loop.

What do we think about the above code ? @hugsy @daviduwu9  (plus an `#include <string.h>` at the top of cause)
This way we check instruction by instruction and can fail accordingly.

NOTE: this code is untested, has probably off-by-one errors, I'm not sure about `strtok_r`; and the variables and `goto`s can probably be done a bit more clever, but I am just brain-stoming in C pretty much :thinking: 

EDIT: also, its probably interesting which bytes mismatch, we could print that out, too.

hugsy commented:
Yes that's not a problem at all, I can update the code now, also make it a bit cleaner too.

dderjoel commented:
Could we put the logic of checking into some file, which we include, and then have a `c` file, which then defines `CODE` with the code we want to check?
like `check_add.c` with 
```c
/* each test file only changes this*/
#define CODE "add rax, rbx\n" \
"add al, 0x01\n"\


/* rest is than copy-n-paste boilerplate */

#include "test_common.c" // implements compare_asmline_cc
__attribute__((noinline, section("mytest"), naked)) void
test(void)
{
    __asm__(CODE);
}

int main(){
char * code = CODE;
compare_asmline_cc(&code, &test);
}
```
Maybe I'm a bit in 'make-a-wish' land now, but we could maybe even put all the boilerplate in a file and let the test runner `cat`them together into a temporary file, which is then run.

My intention to split that is, that otherwise we'd have one crazy large `c` file with all the tests in one string eventually.

Commit:
improved `compile_asm`

hugsy commented:
Done, now it'll display a comprehensible error message like this (note: the following was faked for demo purpose - the debug lines won't show either)

<img width="545" alt="image" src="https://user-images.githubusercontent.com/590234/156481958-ade1379e-cc49-487b-91c0-acfb52faafcc.png">

hugsy commented:
> Could we put the logic of checking into some file, which we include, and then have a c file, which then defines CODE with the code we want to check?

Yup, ~~a shell script (simple sed script would be enough) can generate the C from cmake and have it compiled~~ I misread. Yes I can put those in a header too to split the format.

Commit:
using cmake + compile_asm.c to generate test cases

hugsy commented:
You can check out c200b34 I think that's what you had in mind.

Simply put: we use `cmake` to generate compilable code from the test cases located in `test/inc/`. This way we can convert the `test/*asm` quickly to usable C files and have *all* the test cases preserved, and robustly executed. Let me know if you're ok with that and I will finish the transition of the `*.asm` files to that format.

Cheers,

dderjoel commented:
```bash
mkdir build && cd build
cmake .. -DBUILD_TEST=ON
make -j
```
results in
```
fatal error: test_adc.h: No such file or directory
   19 | #include <test_adc.h>
```
same as here
https://github.com/hugsy/AssemblyLine/runs/5401322133?check_suite_focus=true#step:9:235

Edit: Is the `inc` directory supposed to be created by `cmake` ?

hugsy commented:
Ah ... because of the .gitignore, hold on

Commit:
damn gitignore

Commit:
test disabling compile_asm for msvc

dderjoel commented:
Nice I'm doing
```bash
rm -rf build
mkdir build && cd build
cmake .. -DBUILD_TEST=ON
make -j
cd test && ctest -N
```
and it outputs 
```
Test project /tmp/hugsy/build/test
  Test #1: check_chunk_counting
  Test #2: invalid
  Test #3: jump
  Test #4: memory_reallocation
  Test #5: optimization_disabled
  Test #6: run
  Test #7: vector_operations
```
which is what I like seeing: no failed tests. But It does not seem to run the `compare_asm.c` right?
I also changed `src/instructions.c` to create invalid code for `adc`. And the `ctest -N` still runs fine.
But then also `test/CMakeLists.txt`:17 clearly states `compile_asm.c`.

Am I calling it incorrectly?

hugsy commented:
Yup, `ctest` doesn't apply to those.
Is this approach ok for you? Just to know if I should finish it (because this PR is getting quite long now 😁)

dderjoel commented:
yes, I love it! how would I call them ?

Commit:
move test to new format

hugsy commented:
Ok, I've sent you what I've got so far. The big issue I'm facing now is that you chose to support the nasm syntax which defines more than just respecting the intel abi (instruction mangling, macro definitions, etc.) and those cannot be understood by anything else than ... nasm. So currently the build fails because gcc/clang see those instructions as invalid and therefore cannot compile some test binaries. 

There's no easy solution for me to that because it's an implementation choice. What could be done right now is disregarding those nasm edge cases and only test the real intel abi. And later find a solution for adding tests to those cases.

Please feel free to read the code, I added some comments hoping they'll help. If you have question, let me know.

Cheers,

Commit:
comment the test cases whose syntax is currently not working

Commit:
applied clang-formatting

Commit:
fixed ci

Commit:
fixing last CI tests

dderjoel commented:
Alright.  Thank you so much for your help. We do have another (private) assemblyline develop-repository, which contains a handful of bugfixes / features. I will check with @daviduwu9, how we will merge all those three repos together. 
@daviduwu9 I think we should merge all the (new) test-cases from the dev-reop into a fork of `hugsy:main`, then apply the code-changes from the dev-repo into the fork. Then we should be able to put all of that into this `0xADE1A1DE`-repo. Maybe as  a second branch first, check `CI`, have a beta-release, and if we test this on a couple machines we can merge it into main as version `1.3.0`. As this is a very substantial change in code, we could consider add a couple breaking changes (remove deprecated functions / change directory layout, change `-p` options...) and release `2.0.0`.

Again, @hugsy, Thank you very much I think this is a great contribution. We will come back here if we need assistance or ping, once we are merged to main. 

Cheers!

Commit:
ci fixes

hugsy commented:
No problem. Feel free to close this PR if you prefer.
I'll keep working on it in my fork.

Cheers

Commit:
ci bad windows path
